File: Chapter05/MoveComponent.cpp
//
//  MoveComponent.cpp
//  Game-mac
//
//  Created by Michael Lehmann on 12.01.21.
//  Copyright © 2021 Michael Lehmann. All rights reserved.
//

#include "MoveComponent.h"
#include "Actor.h"
#include "Math.h"

MoveComponent::MoveComponent(class Actor* owner, int updateOrder)
    :Component(owner, updateOrder) { }

void MoveComponent::Update(float deltaTime) {
    if (!Math::NearZero(mAngularSpeed)) {
        float rot = mOwner->GetRotation();
        rot += mAngularSpeed * deltaTime;
    
        mOwner->SetRotation(rot);
    }
    
    if (!Math::NearZero(mForwardSpeed)) {
        Vector2 pos = mOwner->GetPosition();
        pos += mOwner->GetForward() * mForwardSpeed * deltaTime;
        
        // (Screen wrapping code only for asteroids)
        if (pos.x < 0.0f) { pos.x = 1022.0f; }
        else if (pos.x > 1024.0f) { pos.x = 2.0f; }

        if (pos.y < 0.0f) { pos.y = 766.0f; }
        else if (pos.y > 768.0f) { pos.y = 2.0f; }

        mOwner->SetPosition(pos);
    }
}

File: Chapter05_Veldrid/MoveComponent.cs
using System;
using System.Numerics;

namespace Chapter05
{
    public class MoveComponent : Component
    {
        public MoveComponent(Actor owner, int updateOrder = 10)
            : base(owner, updateOrder)
        {
        }

        public float AngularSpeed { get; set; }

        public float ForwardSpeed { get; set; }

        public override void Update(float deltaTime)
        {
            if (!AngularSpeed.NearZero())
            {
                float rotation = Owner.Rotation;
                rotation += AngularSpeed * deltaTime;
                Owner.Rotation = rotation;
            }

            if (!ForwardSpeed.NearZero())
            {
                Vector2 position = Owner.Position;
                position += Owner.Forward * ForwardSpeed * deltaTime;

                // Screen wrapping (for asteroids)
                var halfScreenWidth = Owner.Game.Renderer.Window.Width / 2;
                if (position.X < -halfScreenWidth)
                {
                    position.X = halfScreenWidth - 2;
                }
                else if (position.X > halfScreenWidth)
                {
                    position.X = -(halfScreenWidth - 2);
                }

                var halfScreenHeight = Owner.Game.Renderer.Window.Height / 2;
                if (position.Y < -halfScreenHeight)
                {
                    position.Y = halfScreenHeight - 2;
                }
                else if (position.Y > halfScreenHeight)
                {
                    position.Y = -(halfScreenHeight - 2);
                }

                Owner.Position = position;
            }
        }
    }
}